### ISSUE TITLE
Minor modes

### ISSUE THREAD

**Issue by emdemir** [i3/i3#4158]
## I'm submitting a…
<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

In i3, there is currently only one "mode" construct, where all previous keybinds are removed. This is quite restrictive, as one might want to temporarily switch to a different state to perform a command while still keeping the functionality of their other keybinds, or possibly keeping it on as necessary.

## Desired Behavior

I propose a system similar to Emacs' minor modes. In Emacs, minor modes aid the active major mode by providing additional functionality. In the same sense, an *i3 minor mode* would be a mode which supplements the current configuration by adding or changing keybinds, and perhaps executing commands at enter or exit.

Let me give a few examples. First of all, the syntax for minor modes would be very similar to the mode system that already exists in i3:
```
minor_mode "mode_name" {
# ...statements...
}
```

However, unlike regular modes, minor modes would not remove the already existing keybinds. Multiple minor modes could be entered simultaneously (i3 would perhaps complain if more than one minor mode tried to bind the same key).

So, you could make something like an application launcher without having to assign permanently dedicated key combinations:
```
minor_mode "launcher" {
    bindsym s exec --no-startup-id steam
    bindsym f exec --no-startup-id firefox
    # and so on and so forth
    bindsym Escape minor_mode --off "launcher"
}
```

Additionally, you would be able to execute commands when the mode is entered and when it is exited. This might make it easier to add minor modes which alter your desktop in some way in a temporary fashion. For example, I use the US keyboard in my day-to-day tasks, but I want to sometimes be able to type in Turkish. This would make it really simple to add a simple toggling minor mode:
```
minor_mode "turkish" {
    exec setxkbmap tr
    atexit exec setxkbmap us

    bindsym $mod+space minor_mode --off "turkish"
}
bindsym $mod+space minor_mode "turkish"
```

I also propose a new keyboard that you can see above, `atexit`. This would execute the command after the minor mode is turned off (by a key combination or what have you).

I will try to work on adding such a feature but I don't know how complex this would be or whether I will have enough time to do it. Hope this isn't too outrageous of a request.

**Comment by rpigott** [comment]
What if you could just say `inherit <mode here>` within a mode definition? Modes would retain all the keybinds from the other modes they inherit. Seems easier than introducing a new mode type since it just changes which bindings are defined for the modes, only without you having to copy them all in.

I guess you wouldn't be able to activate more than one at once (unless you also defined some combination modes that included several other modes), but it also sidesteps the issue of what to do when multiple modes define the same keybind, and doesn't require a new concept in i3.

As for executing commands on mode changes, you can already do this by just changing your keybinds to do [both actions](https://i3wm.org/docs/userguide.html#command_chaining). Instead of `bindsym $mod+blah mode turkish`, `bindsym $mod+blah xkbmapblah; mode turkish`. You can also have a client subscribe to the [mode](https://i3wm.org/docs/ipc.html#_mode_event) event.

**Comment by emdemir** [comment]
Yeah, my main point of introducing this new syntax would be to allow multiple modes to be active at once, and enabled/disabled independently. As for conflicts, i3 already complains about conflicting keybinds at the toplevel, just doing the same check during runtime should suffice I think.

Commands can indeed be run like that, yes, but it would either need your `inherit` command.

**Comment by rpigott** [comment]
> As for conflicts, i3 already complains about conflicting keybinds at the toplevel, just doing the same check during runtime should suffice I think.

At least in the case of using `inherit`, I think it would be preferable to allow redefinition of bindings from the inherited mode. That would make it easier to define a mode that replaces a subset of the default bindings.

I understand that `inherit` may not be as "powerful" as your suggestion, but I think it strikes a good balance as something that is similarly useful, but easier to implement. Most importantly, it does not require considering how to extend or modify the ipc (get_binding_modes and get_binding_state queries, binding and mode events, possibly new queries/events, not to mention the swaybar protocol if the bar wants to report active minor modes) to accommodate a new feature, which is unrelated to the actual implementation of the feature, and would be the the real hurdle of trying to add "minor modes".

**Comment by stapelberg** [comment]
As an Emacs user myself, I’m sympathetic to the idea of major/minor modes, but at the same time, I’m worried about making this area of i3 more complex than it absolutely has to be.

Is there any chance that this could be implemented external to i3 using the IPC interface?

**Comment by rpigott** [comment]
> Yeah, my main point of introducing this new syntax would be to allow multiple modes to be active at once, and enabled/disabled independently. As for conflicts, i3 already complains about conflicting keybinds at the toplevel, just doing the same check during runtime should suffice I think.

If we do this, then any minor modes you define can't have common keys bound or they will conflict and you wouldn't be able to activate them both at once anyway. If that restriction is acceptable, it makes `inherit` much more attractive imo since you could reasonably get the same behavior by defining more regular modes

> Is there any chance that this could be implemented external to i3 using the IPC interface?

If it were implemented with the IPC, it would still need a way to actually change the keybinds. I imagine that means defining combination modes, perhaps programmatically, using the regular syntax and then switching between those. I think an `inherit` command would simplify that too.

Any thoughts on a mode `inherit` statement? It's certainly much simpler and I think goes a long way toward satisfying what users might want out of this feature, and could help to implement it in a more feature complete way outside of i3.

**Comment by Airblader** [comment]
Personally I think the added complexity is not justified. Shortcut handling can be managed through external tools, and some are already more powerful than what i3 offers, and I think that's fine. As for something like inherit, that's just syntactic sugar and tools like i3bang probably make that easy enough already.

**Comment by a-pav** [comment]
I support the `inherit` notion. It's indeed bothersome not being able to (for e.g.) switch between workspaces/tabs while you have a non-default mode triggered.
Initially I thought of the idea when I was about to write an _automatic mode switcher_ using the IPC, but with no `inherit` thing, it wouldn't have much use. 
As for i3bang, it's a bit too much and I don't get the syntax!
